# Neural Networks AI

## Exercise 0: Environment and libraries

In [3]:
!python --version

Python 3.11.7


In [4]:
!pip list

Package                   Version
------------------------- ---------------
anyio                     4.2.0
appnope                   0.1.3
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 2.4.1
async-lru                 2.0.4
attrs                     23.2.0
Babel                     2.14.0
beautifulsoup4            4.12.3
bleach                    6.1.0
blinker                   1.7.0
certifi                   2023.11.17
cffi                      1.16.0
charset-normalizer        3.3.2
click                     8.1.7
comm                      0.2.1
contourpy                 1.2.0
cycler                    0.12.1
dash                      2.15.0
dash-core-components      2.0.0
dash-html-components      2.0.0
dash-table                5.0.0
debugpy                   1.8.0
decorator                 5.1.1
defusedxml                0.7.1
executing                 2.0.1
fastjsonschema            2.19.1
Flask         

## Exercise 1: The neuron

### Task Description

The goal of this exercise is to understand the role of a neuron and to implement a neuron.

An artificial neuron, the basic unit of the neural network, (also referred to as a perceptron) is a mathematical function. It takes one or more inputs that are multiplied by values called “weights” and added together. This value is then passed to a non-linear function, known as an activation function, to become the neuron’s output.

As described in the article, **a neuron takes inputs, does some math with them, and produces one output**.

Let us assume there are 2 inputs. Here are the three steps involved in the neuron:

1. Each input is multiplied by a weight
   - x1 -> x1 \* w1
   - x2 -> x2 \* w2
2. The weighted inputs are added together with a biais b
   - (x1 _ w1) + (x2 _ w2) + b
3. The sum is passed through an activation function

   - y = f((x1 _ w1) + (x2 _ w2) + b)

   - The activation function is a function you know from W2DAY2 (Logistic Regression): **the sigmoid**

Example:

x1 = 2 , x2 = 3 , w1 = 0, w2= 1, b = 4

1. Step 1: Multiply by a weight
   - x1 -> 2 \* 0 = 0
   - x2 -> 3 \* 1 = 3
2. Step 2: Add weighted inputs and bias
   - 0 + 3 + 4 = 7
3. Step 3: Activation function
   - y = f(7) = 0.999

**Non-vectorized version**

In [5]:
import math

class Neuron:
    def __init__(self, weight1, weight2, bias):
        self.weight1 = weight1
        self.weight2 = weight2
        self.bias = bias

    def feedforward(self, input1, input2):
        # Multiply inputs with weights and add bias
        total = input1 * self.weight1 + input2 * self.weight2 + self.bias
        # Sigmoid activation function
        return 1 / (1 + math.exp(-total))

#### Audit 1:

In [7]:
neuron = Neuron(0,1,4)
neuron.feedforward(2,3)

0.9990889488055994

**Vectorized version**

In [6]:
import numpy as np

class VectorizedNeuron:
    def __init__(self, weights, bias):
        self.weights = np.array(weights)
        self.bias = bias

    def feedforward(self, inputs):
        # Using dot product for weighted sum
        total = np.dot(self.weights, inputs) + self.bias
        # Sigmoid activation function
        return 1 / (1 + np.exp(-total))

# Create a vectorized neuron with given weights and bias
vectorized_neuron = VectorizedNeuron([0, 1], 4)
# Test the neuron with provided inputs as an array
vectorized_output = vectorized_neuron.feedforward(np.array([2, 3]))
vectorized_output

0.9990889488055994

## Exercise 2: Neural network

The goal of this exercise is to understand how to combine three neurons to form a neural network. A neural network is nothing else than neurons connected together. As shown in the figure the neural network is composed of **layers**:

- Input layer: it only represents input data. **It doesn't contain neurons**.
- Output layer: it represents the last layer. It contains a neuron (in some cases more than 1).
- Hidden layer: any layer between the input (first) layer and output (last) layer. Many hidden layers can be stacked. When there are many hidden layers, the neural networks is deep.

Notice that the neuron **o1** in the output layer takes as input the output of the neurons **h1** and **h2** in the hidden layer.

In exercise 1, you implemented this neuron.

Now, we add two more neurons:

- h2, the second neuron of the hidden layer
- o1, the neuron of the output layer

1. Implement the function `feedforward` of the class `OurNeuralNetwork` that takes as input the input data and returns the output y. Return the output for these neurons:

   ```
   neuron_h1 = Neuron(1,2,-1)
   neuron_h2 = Neuron(0.5,1,0)
   neuron_o1 = Neuron(2,0,1)
   ```

   ```
   class OurNeuralNetwork:

       def __init__(self, neuron_h1, neuron_h2, neuron_o1):
           self.h1 = neuron_h1
           self.h2 = neuron_h2
           self.o1 = neuron_o1

       def feedforward(self, x1, x2):
       # The inputs for o1 are the outputs from h1 and h2
       # TODO
           return y

   ```


In [8]:
class OurNeuralNetwork:
    def __init__(self, neuron_h1, neuron_h2, neuron_o1):
        self.h1 = neuron_h1
        self.h2 = neuron_h2
        self.o1 = neuron_o1

    def feedforward(self, x1, x2):
        # Outputs from the hidden layer neurons
        output_h1 = self.h1.feedforward(x1, x2)
        output_h2 = self.h2.feedforward(x1, x2)

        # Output from the output layer neuron, using outputs of hidden layer as inputs
        y = self.o1.feedforward(output_h1, output_h2)
        return y

# Initialize neurons with given weights and biases
neuron_h1 = Neuron(1, 2, -1)
neuron_h2 = Neuron(0.5, 1, 0)
neuron_o1 = Neuron(2, 0, 1)

# Create an instance of OurNeuralNetwork
neural_network = OurNeuralNetwork(neuron_h1, neuron_h2, neuron_o1)

# Test the feedforward function with some inputs
output = neural_network.feedforward(3, 4)  # Example inputs
output

0.9525700248057429

#### Audit 2

In [10]:
neural_network.feedforward(2,3)

0.9524917424084265

## Exercise 3: Log loss

The objective of this exercise is to implement the Log Loss function, which serves as a **loss function** in classification problems. This function quantifies the difference between predicted and actual categorical outcomes, producing lower values for accurate predictions.

Log Loss is a function used in neural networks to help find the best weights for accurate predictions, similar to how we use Mean Squared Error (MSE) to improve predictions in linear regression. While MSE works well for regression (predicting numbers), Log Loss is specifically designed for classification tasks (predicting categories).

Log Loss is computed using the formula:

`Log loss: - 1/n * Sum[(y_true*log(y_pred) + (1-y_true)\*log(1-y_pred))]`

This equation calculates Log Loss across all predictions in a dataset, penalizing the model more for larger discrepancies between predicted and actual class probabilities.

1.  Create a function `log_loss_custom` and compute the loss for the data below:

        ```
        y_true = np.array([0,1,1,0,1])
        y_pred = np.array([0.1,0.8,0.6, 0.5, 0.3])
        ```
        Check that `log_loss` from `sklearn.metrics` returns the same result

    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.log_loss.html

In [11]:
from sklearn.metrics import log_loss

def log_loss_custom(y_true, y_pred):
    # Calculate the log loss for each pair of true and predicted values
    n = len(y_true)
    loss = -sum(y_true * np.log(y_pred) + (1 - y_true) * np.log(1 - y_pred)) / n
    return loss

# Given data
y_true = np.array([0, 1, 1, 0, 1])
y_pred = np.array([0.1, 0.8, 0.6, 0.5, 0.3])

# Calculate log loss using the custom function
custom_log_loss = log_loss_custom(y_true, y_pred)

# Calculate log loss using sklearn's function for comparison
sklearn_log_loss = log_loss(y_true, y_pred)

custom_log_loss, sklearn_log_loss

(0.5472899351247816, 0.5472899351247816)

#### audit 3

For question 1, is the output 0.5472899351247816?

## Exercise 4: Forward propagation

The goal of this exercise is to compute the log loss on the output of the forward propagation. The data used is the tiny data set below.

| name | math | chemistry | exam_success |
| :--- | ---: | --------: | -----------: |
| Bob  |   12 |        15 |            1 |
| Eli  |   10 |         9 |            0 |
| Tom  |   18 |        18 |            1 |
| Ryan |   13 |        14 |            1 |

The goal if the network is to predict the success at the exam given math and chemistry grades. The inputs are `math` and `chemistry` and the target is `exam_success`.

1. Compute and return the output of the neural network for each of the students. Here are the weights and biases of the neural network:

   ```
   neuron_h1 = Neuron(0.05, 0.001, 0)
   neuron_h2 = Neuron(0.02, 0.003, 0)
   neuron_o1 = Neuron(2,0,0)
   ```

2. Compute the logloss for the data given the output of the neural network with the 4 students.

In [16]:
# Initializing the neurons with the given weights and biases
neuron_h1 = Neuron(0.05, 0.001, 0)
neuron_h2 = Neuron(0.02, 0.003, 0)
neuron_o1 = Neuron(2, 0, 0)

# Creating an instance of the neural network
neural_network = OurNeuralNetwork(neuron_h1, neuron_h2, neuron_o1)

# Students' data
students_data = {
    "Bob": {"math": 12, "chemistry": 15, "exam_success": 1},
    "Eli": {"math": 10, "chemistry": 9, "exam_success": 0},
    "Tom": {"math": 18, "chemistry": 18, "exam_success": 1},
    "Ryan": {"math": 13, "chemistry": 14, "exam_success": 1}
}

# Computing the output for each student
outputs = {name: neural_network.feedforward(data["math"], data["chemistry"]) 
           for name, data in students_data.items()}

outputs

{'Bob': 0.7855253278357536,
 'Eli': 0.7771516558846259,
 'Tom': 0.8067873659804015,
 'Ryan': 0.7892343955586032}

#### Audit 4.1

```
Bob: 0.7855253278357536
Eli: 0.7771516558846259
Tom: 0.8067873659804015
Ryan: 0.7892343955586032

In [17]:
# Extracting the actual labels (exam_success) and predicted probabilities
y_true = np.array([data["exam_success"] for data in students_data.values()])
y_pred = np.array(list(outputs.values()))

# Calculate log loss for the dataset
log_loss_dataset = log_loss_custom(y_true, y_pred)
log_loss_dataset

0.5485133607757963

#### Audit 4.2

is the logloss for the 4 students 0.5485133607757963?

## Exercise 5: Regression

The goal of this exercise is to learn to adapt the output layer to regression.
As a reminder, one of reasons for which the sigmoid is used in classification is because it contracts the output between 0 and 1 which is the expected output range for a probability (W2D2: Logistic regression). However, the output of the regression is not a probability.

In order to perform a regression using a neural network, the activation function of the neuron on the output layer has to be modified to **identity function**. In mathematics, the identity function is: **f(x) = x**. In other words it means that it returns the input as so. The three steps become:

1. Each input is multiplied by a weight
   - x1 -> x1 \* w1
   - x2 -> x2 \* w2
2. The weighted inputs are added together with a biais b
   - (x1 _ w1) + (x2 _ w2) + b
3. The sum is passed through an activation function
   - y = f((x1 _ w1) + (x2 _ w2) + b)
   - The activation function is **the identity**
   - y = (x1 _ w1) + (x2 _ w2) + b

All other neurons' activation function **doesn't change**.

1. Adapt the neuron class implemented in exercise 1. It now takes as a parameter `regression` which is boolean. When its value is `True`, `feedforward` should use the identity function as activation function instead of the sigmoid function.

   ```
   class Neuron:
   def __init__(self, weight1, weight2, bias, regression):
       self.weights_1 = weight1
       self.weights_2 = weight2
       self.bias = bias
       #TODO

   def feedforward(self, x1, x2):
       #TODO
       return y

   ```

   - Compute the output for:

     ```
     neuron = Neuron(0,1,4, True)
     neuron.feedforward(2,3)
     ```

2. Now, the goal of the network is to predict the physics' grade at the exam given math and chemistry grades. The inputs are `math` and `chemistry` and the target is `physics`.

| name | math | chemistry | physics |
| :--- | ---: | --------: | ------: |
| Bob  |   12 |        15 |      16 |
| Eli  |   10 |         9 |      10 |
| Tom  |   18 |        18 |      19 |
| Ryan |   13 |        14 |      16 |

Compute and return the output of the neural network for each of the students. Here are the weights and biases of the neural network:

```
    #replace regression by the right value
    neuron_h1 = Neuron(0.05, 0.001, 0, regression)
    neuron_h2 = Neuron(0.002, 0.003, 0, regression)
    neuron_o1 = Neuron(2,7,10, regression)
```

3. Compute the MSE for the 4 students.


In [19]:
import numpy as np

class Neuron:
    def __init__(self, weight1, weight2, bias, regression=False):
        self.weight1 = weight1
        self.weight2 = weight2
        self.bias = bias
        self.regression = regression

    def feedforward(self, x1, x2):
        total = x1 * self.weight1 + x2 * self.weight2 + self.bias
        if self.regression:
            # For regression, use the identity function
            return total
        else:
            # For classification, use the sigmoid function
            return 1 / (1 + np.exp(-total))

class OurNeuralNetwork:
    def __init__(self, neuron_h1, neuron_h2, neuron_o1):
        self.h1 = neuron_h1
        self.h2 = neuron_h2
        self.o1 = neuron_o1

    def feedforward(self, x1, x2):
        output_h1 = self.h1.feedforward(x1, x2)
        output_h2 = self.h2.feedforward(x1, x2)
        return self.o1.feedforward(output_h1, output_h2)

# Initialize neurons, only output neuron is for regression
neuron_h1 = Neuron(0.05, 0.001, 0, False)
neuron_h2 = Neuron(0.002, 0.003, 0, False)
neuron_o1 = Neuron(2, 7, 10, True)

# Neural network instance
neural_network = OurNeuralNetwork(neuron_h1, neuron_h2, neuron_o1)

# Students' data
students_data = {
    "Bob": {"math": 12, "chemistry": 15, "physics": 16},
    "Eli": {"math": 10, "chemistry": 9, "physics": 10},
    "Tom": {"math": 18, "chemistry": 18, "physics": 19},
    "Ryan": {"math": 13, "chemistry": 14, "physics": 16}
}

# Compute outputs and MSE
outputs = {name: neural_network.feedforward(data["math"], data["chemistry"]) 
           for name, data in students_data.items()}

y_true = np.array([data["physics"] for data in students_data.values()])
y_pred = np.array(list(outputs.values()))
mse = np.mean((y_true - y_pred)**2)

print("Outputs:", outputs)
print("MSE:", mse)

Outputs: {'Bob': 14.918863163724454, 'Eli': 14.83137890625537, 'Tom': 15.086662606964074, 'Ryan': 14.939270885974128}
MSE: 10.237608699909138


In [20]:
import numpy as np

class Neuron:
    def __init__(self, weight1, weight2, bias, regression=False):
        self.weight1 = weight1
        self.weight2 = weight2
        self.bias = bias
        self.regression = regression

    def feedforward(self, x1, x2):
        total = x1 * self.weight1 + x2 * self.weight2 + self.bias
        if self.regression:
            # For regression, use the identity function
            return total
        else:
            # For classification, use the sigmoid function
            return 1 / (1 + np.exp(-total))

class OurNeuralNetwork:
    def __init__(self, neuron_h1, neuron_h2, neuron_o1):
        self.h1 = neuron_h1
        self.h2 = neuron_h2
        self.o1 = neuron_o1

    def feedforward(self, x1, x2):
        output_h1 = self.h1.feedforward(x1, x2)
        output_h2 = self.h2.feedforward(x1, x2)
        return self.o1.feedforward(output_h1, output_h2)

# Test case for Neuron class
neuron_test = Neuron(0, 1, 4, True)
test_output = neuron_test.feedforward(2, 3)
print("Test Output:", test_output)  # Expected to be 7

# Initialize neurons, only output neuron is for regression
neuron_h1 = Neuron(0.05, 0.001, 0, False)
neuron_h2 = Neuron(0.002, 0.003, 0, False)
neuron_o1 = Neuron(2, 7, 10, True)

# Neural network instance
neural_network = OurNeuralNetwork(neuron_h1, neuron_h2, neuron_o1)

# Students' data
students_data = {
    "Bob": {"math": 12, "chemistry": 15, "physics": 16},
    "Eli": {"math": 10, "chemistry": 9, "physics": 10},
    "Tom": {"math": 18, "chemistry": 18, "physics": 19},
    "Ryan": {"math": 13, "chemistry": 14, "physics": 16}
}

# Compute outputs and MSE
outputs = {name: neural_network.feedforward(data["math"], data["chemistry"]) 
           for name, data in students_data.items()}

y_true = np.array([data["physics"] for data in students_data.values()])
y_pred = np.array(list(outputs.values()))
mse = np.mean((y_true - y_pred)**2)

print("Outputs:", outputs)
print("MSE:", mse)

Test Output: 7
Outputs: {'Bob': 14.918863163724454, 'Eli': 14.83137890625537, 'Tom': 15.086662606964074, 'Ryan': 14.939270885974128}
MSE: 10.237608699909138


#### Audit 5

For question 1, is the output 7?
For question 2, are the outputs the following?
```
Bob: 14.918863163724454
Eli: 14.83137890625537
Tom: 15.086662606964074
Ryan: 14.939270885974128
```
For question 3, is the MSE 10.237608699909138?